torch/distributed/checkpoint/planner.py

```diff
@@ -426,3 +427,31 @@ def commit_tensor(self, read_item: ReadItem, tensor: torch.Tensor) -> None:
         The contents of tensor will follow its device synchronization model.
         """
         pass
+
+
+class Checkpointable:
+    """
+    Interface for checkpointable objects.
+    This is to allow arbitrary objects/tensor subclasses to hook into DCP seamlessly through implementing the interface.
+    """
+
+    @abc.abstractmethod
+    def _create_write_items(self, fqn: str, object: Any) -> List[WriteItem]:
+        """
+        Return a list of WriteItems based on object's contents.
+        """
+        raise NotImplementedError("Checkpointable._create_write_items is not implemented")
+
+    @abc.abstractmethod
+    def _create_chunk_list(self, tensor: torch.Tensor) -> List[ChunkStorageMetadata]:
+        """
+        Return a list of `ChunkStorageMetadata` based on object's contents.
+        """
+        raise NotImplementedError("Checkpointable._create_chunk_list is not implemented")
+
+    @abc.abstractmethod
+    def find_tensor_shard(self, tensor: torch.Tensor, index: MetadataIndex) -> torch.Tensor:
```

Comment:
can we rename this as "_get_tensor_shard"? so that it's not a public facing API?